### ISSUE TITLE
System.Net.HttpListener make memory leak by .net core only.

### ISSUE THREAD

**Issue by moonmile** [dotnet/corefx#32454]
This is test code.

```c#
namespace ConsoleApp2
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("test server");
            var listener = new System.Net.HttpListener();
            listener.Prefixes.Add("http://127.0.0.1:5411/");
            listener.Start();
            while (true)
            {
                var context = listener.GetContext();
                var res = context.Response;
                res.StatusCode = 200;
                var sw = new System.IO.StreamWriter(res.OutputStream);
                sw.Write(string.Format("text {0}", DateTime.Now.ToString()));
                sw.Close();
            }
        }
    }
}
```
When hundreds of times to access this simple HTTP server, this code make memory leak.
When I compiled by .net core and execute, the memory consumed little by little will increase.
But it is right memory, when I compiled by .NET Framework 4.6.

**Comment by moonmile** [comment]
pattern of .NET Framework 4.6
![image](https://user-images.githubusercontent.com/1286584/46023210-242ee480-c11f-11e8-86ed-c8ac795e7db5.png)

pattern of .NET Core 2.1
![image](https://user-images.githubusercontent.com/1286584/46023254-37da4b00-c11f-11e8-8e72-d3b84849dc8a.png)

When .NET Framework 6.4, used memory reaches the limit in a few seconds.
But, when .NET Core 2.1,  used memory keeps increasing.

**Comment by jbhensley** [comment]
Reproducible using supplied code. Just pop the following in PowerShell for client requests:

```PowerShell
for($i = 0;$i -lt 10000;$i++){
    Invoke-WebRequest 127.0.0.1:5411
    $i
}
```

Memory on .NET Framework 4.7.1 stabilizes after about 500 iterations or so. Core 2.1 keeps mountain climbing.

**Comment by jbhensley** [comment]
The odd thing is that memory snapshots are not showing any increase in object instance count or heap size:

![memory diagnostics](https://user-images.githubusercontent.com/13544395/46044757-84e30080-c0d0-11e8-8277-ecb4b743fc57.png)

The first snapshot was just after launching the app, but before any client requests. There was a burst of object creation after starting the request loop, but nothing after that. Is process memory registering total process mem size, but snapshot is only showing managed objects? A leak in native memory perhaps?

**Comment by jbhensley** [comment]
Confirmed to be in native memory. Use instructions from [here](https://github.com/dotnet/project-system/issues/1125#issuecomment-349163889) to get native debugging working in core. Sorry for piece-milling this.

![native](https://user-images.githubusercontent.com/13544395/46049501-a39ec280-c0e3-11e8-9492-df6dce3ed49d.png)

![void](https://user-images.githubusercontent.com/13544395/46049610-350e3480-c0e4-11e8-8fa5-91149fe5e28a.png)

**Comment by moonmile** [comment]
Thank for your check.

I tried to go this client code for check.
if .net framework 4.6, this memory reaches the upper limit after a certain amount of time.
if .net core 2.1, this server consume memory as much as possible and crashed.
```
    class Program
    {
        static void Main(string[] args)
        {
            HttpClient client = new HttpClient();
            int cnt = 0;
            while ( true )
            {
                var res = client.GetAsync("http://localhost:5411").Result;
                var text = res.Content.ReadAsStringAsync().Result;
                Console.WriteLine($"{cnt} {text}");
                cnt++;
                System.Threading.Thread.Sleep(20);
            }
        }
    }
```

I lookup this when I make Scratch extension code.
https://github.com/moonmile/FireScratch
I make extension server by HttpListener, and go this program for one hour.
This extension code is crashed. That code is written by F#, but same crashed by C#.

**Comment by jbhensley** [comment]
I just used the PowerShell script to generate client requests because it was quick and easy.

Ok, so I've got a workaround for you and some information that will hopefully help the next person who looks at this be able to track it down. I tried, but I'm just not seeing exactly where it's all going wrong even though it feels like I'm close.

The workaround is actually pretty simple: just close/dispose the response:

```C#
var buffer = new byte[] { (byte)'a' };

var listener = new HttpListener();
listener.Prefixes.Add("http://127.0.0.1:5411/");
listener.Start();

Console.WriteLine("Listening...");
while (true)
{
	var context = listener.GetContext();
	var res = context.Response;
	res.StatusCode = 200;
	res.OutputStream.Write(buffer, 0, buffer.Length);
	res.OutputStream.Flush();
	res.OutputStream.Dispose();
	res.Close();    // the magic
}
```
This could be done with a `using` to make things easier. Why this works, though, is a bit more complicated. It's actually a chain of events:

- `HttpListenerResponse.Close` calls `HttpListenerResponse.Dispose`
- `HttpListenerResponse.Dispose` calls the `HttpListenerContext`'s internally scoped `Close` method
- `HttpListenerContext.Close` calls `HttpListenerRequest.Close`
- `HttpListenerRequest.Close` calls `RequestContextBase.Close` on the request
- `RequestContextBase.Close` calls `RequestContextBase.Dispose`
- `RequestContextBase.Dispose` contains the line that frees native memory: `Marshal.FreeHGlobal(_backingBuffer);`

Breaking any link in that call chain results in a leak. Comparing corefx code to .NET Framework, however, I'm not finding the spot where the framework is releasing and core is not. Granted, it's a complex web. Hopefully someone from the corefx team with more intimate knowledge of this area will be able to take a look.

And, just for reference, here's the memory profile with and without `res.Close()`:

![leak](https://user-images.githubusercontent.com/13544395/46115294-2b9bcf80-c1ab-11e8-8d21-4dd70d73c335.png)

![no leak](https://user-images.githubusercontent.com/13544395/46115298-2fc7ed00-c1ab-11e8-8f1b-1bababb50d6c.png)

**Comment by moonmile** [comment]
I see.
When I make code to use 'res.close' or 'using', its program don't leak memory by .NET Core.

```
        static void Main(string[] args)
        {
            Console.WriteLine("test server");
            var listener = new System.Net.HttpListener();
            listener.Prefixes.Add("http://127.0.0.1:5411/");
            listener.Start();
            while (true)
            {
                var context = listener.GetContext();
                using (var res = context.Response)
                {
                    res.StatusCode = 200;
                    using (var sw = new System.IO.StreamWriter(res.OutputStream))
                    {
                        sw.Write(string.Format("text {0}", DateTime.Now.ToString()));
                    }
                }
			}
		}
```
![image](https://user-images.githubusercontent.com/1286584/46335483-95f8b800-c662-11e8-9b90-18db591cd6f2.png)

**Comment by GSPP** [comment]
Isn't disposing the response object contractually required? Not doing that is simply an API usage error.

Try adding `GC.Collect(); GC.WaitForPendingFinalizers();` to the test loop. There should be no memory leaks because the finalizer should clean up the resources.

Maybe the difference between the runtimes is just different GC behavior.

**Comment by jbhensley** [comment]
I don't think we should be leaking native mem regardless of what happens.

Started looking at this again yesterday. Tried GC'ing the loop, but the finalizer is being suppressed. Whether it gets re-registered currently has no dependency on the state of this particular native allocation, which was a later addition. I am looking to open a PR that removes the suppression, which will effectively solve the problem. However, my concern is that I don't fully understand why suppression was there in the first place. If some chunk of unmanaged mem gets handed off to a native method and we free it in our finalizer while native still has a reference to it, that's going to be a problem.

**Comment by davidsh** [comment]
Fixed in master with #32584.
